packages/cardpay-sdk/sdk/scheduled-payment-module.ts

```diff
@@ -0,0 +1,147 @@
+import { BN } from 'bn.js';
+import { ERC20ABI } from '..';
+import { ZERO_ADDRESS } from './constants';
+
+export default class ScheduledPaymentModule {
+  constructor(private layer2Web3: Web3, private layer2Signer?: Signer) {}
+
+  async enableModule(
+    safeAddress: string,
+    gasTokenAddress: string,
+    txnOptions?: TransactionOptions,
+    contractOptions?: ContractOptions
+  ) {
+    if (!safeAddress) {
+      throw new Error('safeAddress must be specified');
+    }
+    let { nonce, onNonce, onTxnHash } = txnOptions ?? {};
+    let from = contractOptions?.from ?? (await this.layer2Web3.eth.getAccounts())[0];
+
+    let enableModuleTxs = await this.generateEnableModuleTxs(safeAddress);
+    let setGuardTxs = await this.generateSetGuardTxs(safeAddress);
+
+    let multiSendTransaction = await encodeMultiSend(this.layer2Web3, [...enableModuleTxs.txs, ...setGuardTxs.txs]);
+
+    let estimate = await gasEstimate(
+      this.layer2Web3,
+      safeAddress,
+      multiSendTransaction.to,
+      multiSendTransaction.value,
+      multiSendTransaction.data,
+      multiSendTransaction.operation,
+      gasTokenAddress
+    );
+    let gasCost = new BN(estimate.safeTxGas).add(new BN(estimate.baseGas)).mul(new BN(estimate.gasPrice));
+
+    let token = new this.layer2Web3.eth.Contract(ERC20ABI as AbiItem[], gasTokenAddress);
+    let symbol = await token.methods.symbol().call();
+    let balance = new BN(await token.methods.balanceOf(safeAddress).call());
+    if (balance.lt(gasCost)) {
+      throw new Error(
+        `Safe does not have enough balance to enable scheduled payment module. The gas token ${gasTokenAddress} balance of the safe ${safeAddress} is ${fromWei(
+          balance
+        )}, the the gas cost is ${fromWei(gasCost)} ${symbol}`
+      );
+    }
+    if (nonce == null) {
+      nonce = getNextNonceFromEstimate(estimate);
+      if (typeof onNonce === 'function') {
+        onNonce(nonce);
+      }
+    }
+    let signatures = await signSafeTxAsRSV(
+      this.layer2Web3,
+      multiSendTransaction.to,
+      0,
+      multiSendTransaction.data,
+      multiSendTransaction.operation,
+      estimate.safeTxGas,
+      estimate.baseGas,
+      estimate.gasPrice,
+      estimate.gasToken,
+      ZERO_ADDRESS,
+      nonce,
+      from,
+      safeAddress,
+      this.layer2Signer
+    );
+    let gnosisTxn = await executeTransaction(
+      this.layer2Web3,
+      safeAddress,
+      multiSendTransaction.to,
+      multiSendTransaction.data,
+      multiSendTransaction.operation,
+      estimate,
+      nonce,
+      signatures
+    );
+
+    if (typeof onTxnHash === 'function') {
+      await onTxnHash(gnosisTxn.ethereumTx.txHash);
+    }
+
+    return {
+      scheduledPaymentModuleAddress: enableModuleTxs.expectedModuleAddress,
+      metaGuardAddress: setGuardTxs.expectedModuleAddress,
+      txnReceipt: await waitForTransactionConsistency(this.layer2Web3, gnosisTxn.ethereumTx.txHash, safeAddress, nonce),
```

Comment:
so based on our protocol discussion, it sounds like there is client communication to the server to enable a scheduled payment. Perhaps, the client sends the hub the all the parameters that we are currently sending to `executeTransaction`, and the hub actulaly executes the transaction and waits for it, during which it updates it's DB state, and then returns the the client at this point, so that the promise includes all the hub communication as well as txn mining. One argument, though, might be that it's not feasible for the web server to hold on to a client request for the amount of time that it takes to mine the transaction--which likely will be many minutes in ethereum mainnet. That is simply too long for a web server to hold on to a connection. Rather the hub could return the txn hash to the client, and the client can just use web3 to wait for the txn to be mined. @lukemelia @jurgenwerk  it would be good to get your thoughts here too.

Reply:
@habdelra you're right, that's how we specified the general mechanics in the RFC - the client sends params, and waits for the hub to do "everything": executing the transaction, waiting for the transaction to be mined, and returning the result. 

I remember our talk with @alex-cardstack and @FadhlanR where we were discussing the feasibility of giving a single http request so much work and I believe we didn't anticipate these transactions to be taking more than a couple of seconds (maybe we were thinking this is going to be on gnosis chain only?) - I believe that's why we opted for simplifying this by wrapping all the work in a single request. 

In case this can take several minutes then this surely needs to be done async. I'm not sure giving the client the transaction hash and let it wait for it to resolve is the right way to go since our hub is planned to keep track of the transactions so that it can know whether they succeeded or not which then tells whether the scheduled payment has been registered on the contract successfully. One alternative option is to wrap this into a background job in the hub and let the client poll for the job status (the same we did recently for profile creation feature)

But we're talking about creating/canceling a scheduled payment right? Which is a feature separate from this PR which doesn't include any Hub communication as far as I can see

Reply:
@jurgenwerk do we need to store mapping data of user address, safe address, and module address in the crank? If yes, I think it's better we use the same pattern as scheduling/canceling for enabling module.

Reply:
Currently, the eoa, the safe and module address are saved in the scheduled payment row in the DB. When client creates a scheduled payment it sends all these addresses together with the rest of the params and all this gets saved into the scheduled payments table. So we might not need additional actions to maintain that mapping of addresses

Reply:
background job is a nice idea

Reply:
Based on discussions after the daily meeting, we decided to execute the transaction in the SDK, since the hub would have nothing to do with the enabling process, and I removed `await waitForConsistency` because we don't need it here.

Reply:
@FadhlanR the hub doesn't have anything to do with enabling--but it still needs to know that there are schedule payments and populate the DB with the schedule info, so there is definitely communication required between the client and the hub as part of the establishing a scheuled payment. I would think that it would be more reliable to have the server broker the on-chain txn and grab from it any relavant details that it needs.

Reply:
@habdelra I agree to schedule a payment there must be communication between the client and the hub, and the `executeTransaction` must be triggered by the hub. And I also agree we can't rely on the web client to wait for transaction status. But this function is to enable the scheduled payment module, not to schedule payment, So I think we can execute the transaction in SDK and then send a callback via onTxHash callback function and let client do anything with txHash such as send it to the server broker to retrieve the transaction status as you said or just return the etherscan link to the user.